mongodb/snooty
# DOP-4581: Perform client-side redirect based on browser language

### Stories/Links:

DOP-4581

### Current Behavior:

[Prod](https://mongodb.com/docs/atlas)

### Staging Links:

[English docs site example](https://docs-mongodb-org-stg.s3.us-east-2.amazonaws.com/script/master/cloud-docs/raymund.rodriguez/DOP-4581/index.html)
[Portuguese docs site example](https://docs-mongodb-org-stg.s3.us-east-2.amazonaws.com/pt-br/script/master/cloud-docs/raymund.rodriguez/DOP-4581/index.html)

### Notes:

* This "redirect" is client-side logic that changes the `window.href` after the page loads based on the user's browser language settings. This does have the limitation of needing to wait for the JS to load before being sent to the localized page. Ideally, this would be logic on the server side, but since something like that would require changing our existing hosting infrastructure, we opted for client-side for now (see ticket for more information). Note that no 300 response is provided.
* There are two main functionalities introduced in this PR:
   * A user is redirected from the **English** docs site, if a browser language that takes precedence over English is found. The redirect does not happen on non-English docs sites. 
   * A user who has been redirected can use the footer's locale selector to switch back to English to avoid the redirect from happening again (unless the browser's local storage is altered/removed).
* The script performing the redirect is imported to the `head` tag of the HTML so that it can run almost immediately on page load, rather than needing to wait for React hydration if it were used as a typical hook within a component. This does have the caveat of potentially having duplicate code since the script itself cannot use imports due to how Webpack bundles things.
  * For a performance comparsion, compare the above staging links with the following staging links using a React hook: [English docs site example](https://docs-mongodb-org-stg.s3.us-east-2.amazonaws.com/master/cloud-docs/raymund.rodriguez/DOP-4581/index.html) and [Portuguese docs site example](https://docs-mongodb-org-stg.s3.us-east-2.amazonaws.com/pt-br/master/cloud-docs/raymund.rodriguez/DOP-4581/index.html)

**To test:**

1) Go to your browser's language settings and add a language or 12. Rearrange the languages based on desired precedence.
2) Go to the **English** docs staging link. After page render, the URL should change to the localized path of the page. The redirect should only apply to the first non-English language that is compatible with the footer's locale selector.
3) Go to the **Portuguese** docs staging link. You should remain on the Portuguese site regardless of precedent language.
4) On the **Portuguese** docs staging link, use the footer's locale selector to go back to the English docs site.
5) Go to the **English** docs staging link and/or refresh the page. The page should not redirect.

Feel free to remove the new `preferredLocale` key from the `mongodb-docs` object in your browser's local storage if you need to reset settings.

**Open questions**
* How will this impact SEO? Will INP scores be lower due to users being redirected? If this has a negative impact, we should revert this commit, rerelease, and reconsider the viability of using a server-side approach.

### README updates

- - [ ] This PR introduces changes that should be reflected in the README, and I have made those updates.
- - [x] This PR does not introduce changes that should be reflected in the README
